### ISSUE TITLE
Curious about project

### ISSUE THREAD

**Issue by stefnotch** [mgreminger/EngineeringPaper#1]
Just based on the project's title and your GitHub profile, I'd guess that you're trying to build something that combines mathlive and sympy (using pyodide). Is that correct?

If yes, I'd be greatly interested in this project!

**Comment by mgreminger** [comment]
Yes, it's at the initial prototyping phase right now but I do have mathlive, pyodide, and sympy all working together. Just made a couple of commits to the repo that adds the fonts and pyodide files so that every is there to make it run. Eventually want the program to do symbolic equation solving and dimensional analysis using sympy.

**Comment by stefnotch** [comment]
That sounds really great! I'm very much looking forward to seeing how your project develops.

In fact, I had a similar idea a few months back when I realized that I couldn't justify buying MathCad to myself, due to its insane price tag. In case you're curious, [here is the WIP repository with my code](https://github.com/stefnotch/quantum-sheet) and it's [roadmap](https://github.com/stefnotch/quantum-sheet/projects/1). Currently I'm implementing "reactive" variables.
![ezgif-1-daeab246c736](https://user-images.githubusercontent.com/10220080/90308603-146e4400-dee1-11ea-8b49-1ecc7135dc13.gif)


Now, since this is a quite big undertaking, I would like to ask you whether we could join forces and if I could contribute to your repository?

**Comment by mgreminger** [comment]
Yes, it does appear as though we have very similar goals on our projects. There is definitely is a need for opensource software with these capabilities. It's nice to see we've honed in on some of the same technologies to use. Yes, contributions are welcome. Things aren't very stable at this point as I test out different ideas and technologies so it may be a bit of a moving target. I've finally gotten around to updating the [Readme](https://github.com/mgreminger/EngineeringPaper#readme) on the github repo. I've added an animation and a link to a live demo of the current state of the project.

**Comment by stefnotch** [comment]
Great! And yes, I fully agree with you. Especially that we happened to use very similar technologies was a really nice surprise for me.

It's quite understandable that things aren't very stable while it's in the prototyping phase. Mine is also still very much in the prototyping phase. I'll try to get a demo of it online soon.

One thing I'd quite like to do while we're in the prototyping stage, would be to exchange ideas and plans regarding the software. Should we do this over GitHub or is there a different place (Discord, Matrix, Gitter, ...) that you'd recommend?

**Comment by stefnotch** [comment]
I [published it to GitHub](https://stefnotch.github.io/quantum-sheet/), but the most important part (connection with the CAS) isn't in place yet. I'm debating whether I should wait for https://github.com/arnog/mathlive/issues/500 or if I should implement a temporary, possibly regex-based workaround for that.

**Comment by mgreminger** [comment]
The proposed json representation for Mathlive looks very promising and I'll likely switch to that once it's available. Thanks for pointing it out. In the mean time, I've been using the Sympy [`parse_latex`](https://docs.sympy.org/latest/modules/parsing.html#mathrm-latex-parsing-functions-reference) function to parse the latex syntax directly. This, of course, has some limitations due to ambiguous syntax and sympy does not recognize all of the latex syntax used by mathlive (I have a [fork](https://github.com/mgreminger/sympy) to handle some of these cases, eventually I'll create a pull request to get these back into Sympy). `parse_latex` depends on the Python antlr4 runtime, which is not currently part of pyodide, but this is easy to add since it's a pure python library.

**Comment by stefnotch** [comment]
A little update from my side: I tried out the experimental MathJson (https://github.com/arnog/mathlive/issues/500) and it works surprisingly well! I can already type in an expression, convert it to MathJson, convert that to Python code, send the Python code to a web worker and get the result back!

![image](https://user-images.githubusercontent.com/10220080/91427924-9ace4280-e85e-11ea-8863-1933d204b4a8.png)

The next step would be implementing a custom Sympy printer that prints out MathJson. 

By the way, how are you handling the rather large size of pyodide? Are you compiling it yourself?

**Comment by mgreminger** [comment]
Glad to hear that the new MathJson functionality is working well. 

For pyodide, I am compiling it myself. The main reason is because I need to use the development version of sympy with this fix to the dimensional analysis functionality ([sympy #19705](https://github.com/sympy/sympy/pull/19705)) and the antlr4 runtime, which isn't included in pyodide. Since both of these libraries are pure python they could be loaded at runtime as outlined here [pyodide #643](https://github.com/iodide-project/pyodide/issues/643). However, I prefer to have them pre-bundled so that I don't need the micropip dependency.

**Comment by stefnotch** [comment]
Woah, nice commit to Sympy! Glad to know that using Sympy's built-in dimensional analysis is now going to be a bit more bug free. 
I also think that using bundled and if possible, stripped versions of all libraries instead of using micropip to load them at runtime is a better choice.

By the way, Pyodide takes a while to load. Do you have any ideas regarding how to get the loading time down?
(I was thinking about cheating with a SharedWorker that simply keeps running when I reload the page.)

**Comment by mgreminger** [comment]
I'm resigned to the fact that pyodide will take a while to load, I don't think there is a lot ofopportunity to get it much smaller. It's actually quite impressive how small they've gotten it. I do need to add some feedback to the user during the initial load so they are not just faced with an unresponsive app.

**Comment by stefnotch** [comment]
I've done some digging and it seems like a big part of the slowdown is simply due to the size of pyodide. Therefore, this pyodide issue is probably the most relevant one https://github.com/iodide-project/pyodide/issues/646

**Comment by stefnotch** [comment]
Another update from my side: I've implemented a basic Sympy to MathJson printer and hooked it up! It now works quite well.


![ezgif-7-811d552e2607](https://user-images.githubusercontent.com/10220080/91986859-8a6e0a00-ed2d-11ea-9331-12335c476ee7.gif)

(Save for the fact that I only implemented addition and multiplication. I'm waiting for https://github.com/arnog/mathlive/issues/500#issuecomment-683797546)

**Comment by mgreminger** [comment]
Nice work, looks like you're making good progress

**Comment by stefnotch** [comment]
Thank you. How is your prototype coming along? Or, alternatively, what features would you definitely still like to implement before considering the prototype to be done?

**Comment by mgreminger** [comment]
Still quite a few things to work out. Multi-character parameter names and choosing the units for the output (rather than always showing SI units) are probably next on the list.

**Comment by stefnotch** [comment]
Regarding multi-character names, they are a bit tricky with the new mathlive MathJson parser. You can define variables and functions that have multiple characters (e.g. `sin`) and those will get parsed perfectly. However, everything else will get parsed as individual characters (e.g. `ab` gets parsed as `a*b`) 

So, here I wanted to ask you in which cases you'd expect multiple characters to be parsed as a single name. 
e.g.
`myVariable := 7`
`[variableA, variableB] := 9`
`unknownFunction(11, 4)`

**Comment by mgreminger** [comment]
I've been looking a little bit into how to do this with mathlive. There's the same issue with the ambiguity with multiplication with parsing the latex as well. One potential solution is to create an [inline shortcut](https://github.com/arnog/mathlive/tree/master/examples/shortcuts) for the parameter names. Typing VariableName could then be mapped to \VariableName or \mathit{VariableName} to avoid the multiplication ambiguity.

**Comment by stefnotch** [comment]
Yes, that should work quite well. A much bigger problem is when the variable isn't already known. 

For example, here, the variable `a` is never defined. Now, replace `a` with a longer variable name and it's quite ambiguous.
![image](https://user-images.githubusercontent.com/10220080/92525005-5dd15b00-f223-11ea-9932-54b0d8945c0e.png)
(Yes, I recently got symbolic solving to work in my prototype)

Similarly, when defining a new variable, it isn't known yet, but can still have a multi-character name.
![image](https://user-images.githubusercontent.com/10220080/92525291-d7694900-f223-11ea-93f8-326c5de85a2f.png)

**Comment by stefnotch** [comment]
I wrote a bit more about it on the mathlive Gitter, because I think it'd might be quite useful to get the developer's take on this
https://gitter.im/cortex-js/community

**Comment by mgreminger** [comment]
Sounds good, thanks for the link to the discussion.

**Comment by stefnotch** [comment]
By the way, one thing that might be interesting to investigate would be making it possible to have this sort of thing run in a Visual Studio (or other) Notebook. That way, some of the more tedious parts of the UI would already be solved, students could trivially add text explaining what they're doing and it would honestly be really useful to have equations, code, graphs and more in the same notebook.

What do you think?

Screenshot of a notebook:
![screenshot](https://code.visualstudio.com/assets/docs/python/jupyter/native-markdown-03.png)

**Comment by mgreminger** [comment]
Yes, I had briefly looked at that approach, but I'm going for a different execution model than what is supported by the notebooks  (a reactive based approach instead of notebook model where execution order impacts results). Plus I want something that runs entirely within the browser that doesn't require a server (either local or remote).

**Comment by stefnotch** [comment]
I see, that makes sense. 

I personally am actually also eyeing a fully browser based approach, possibly with some notebooks like [Starboard](https://github.com/gzuidhof/starboard-notebook) or [Observable](https://observablehq.com/) or hopefully eventually with [Visual Studio Code's Monaco](https://microsoft.github.io/monaco-editor/). I'll report back if I end up trying out and liking any of the approaches.

I'll definitely keep the reactivity in mind, since after thinking about it, it does seem like something you'd almost expect of a mathematics editor.

By the way, regarding parsing LaTeX, there is [MathJson](https://github.com/cortex-js/math-json#mathjson) which is actively being worked on.
